lib/crypto.js

```diff
@@ -652,6 +672,23 @@ function filterDuplicates(names) {
   }).sort();
 }
 
+// timingSafeEqual reuses the timing-safe Hmac.equals function (see above) for
+// comparison of non-hash inputs, like capability URLs or authentication tokens.
+// Note that the strength of the double-HMAC construction does not depend
+// on the unpredictability of the key. Since randomBytes can potentially
+// throw an exception, we use Math.random, which sacrifices nothing in timing
+// safety and benefits from never throwing an exception.
+// See discussion at
+// https://github.com/nodejs/node-v0.x-archive/issues/8560#issuecomment-60145816
+exports.timingSafeEqual = function(a, b) {
+  var key = new Buffer(32);
+  for (var i = 0; i < 32; i += 4) {
+    key.writeFloatBE(Math.random(), i);
```

Comment:
> timingSafeEqual can still throw

Ah, interesting! Under what conditions?

I was going off the fact that `randomBytes` documents the possibility of an
exception, and was assuming we'd have to similarly cover the possibility
here, even if it was quite rare. All told, I'd prefer to use
`crypto.randomBytes` to avoid the code smell. For me it's mainly a question
of what you'd like the Node API to look like. Is a tiny risk of exception
acceptable? I think, fwiw, that exception can never happen on Linux,
assuming OpenSSL is using /dev/urandom.
On Sep 26, 2015 13:52, "Brendan Ashworth" <notifications@github.com> wrote:

> In lib/crypto.js
> <https://github.com/nodejs/node/pull/3073#discussion_r40498240>:
>
> > @@ -652,6 +672,23 @@ function filterDuplicates(names) {
> >    }).sort();
> >  }
> >
> > +// timingSafeEqual reuses the timing-safe Hmac.equals function (see above) for
> > +// comparison of non-hash inputs, like capability URLs or authentication tokens.
> > +// Note that the strength of the double-HMAC construction does not depend
> > +// on the unpredictability of the key. Since randomBytes can potentially
> > +// throw an exception, we use Math.random, which sacrifices nothing in timing
> > +// safety and benefits from never throwing an exception.
> > +// See discussion at
> > +// https://github.com/nodejs/node-v0.x-archive/issues/8560#issuecomment-60145816
> > +exports.timingSafeEqual = function(a, b) {
> > +  var key = new Buffer(32);
> > +  for (var i = 0; i < 32; i += 4) {
> > +    key.writeFloatBE(Math.random(), i);
>
> timingSafeEqual can still throw, even without crypto.randomBytes (which
> I'm not even sure if that throws regularly). Next-best introduces a
> possible attack vector through Math.random.
>
> That is interesting, about the strength of this. Ultimately it is up to
> you whether or not you want to change it. Perhaps try a benchmark and see
> the difference?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nodejs/node/pull/3073/files#r40498240>.
>